### PR TITLE
CRT Onboarding

### PR DESCRIPTION
**Pull request by mdeggies**
**Onboard Packer OSS to Common Release Tooling**

There are a few important changes to call out in this PR:

- GitHub Actions is being used as the CI provider to build the go binary artifacts, docker image, and linux packages. This logic lives in .`github/workflows/build.yml` and workflow runs are available under https://github.com/hashicorp/packer/actions. 
- Artifact builds will continue to run on PR's and merges to `master`. This behavior can be changed by updating the `on:` logic in the `.github/workflows/build.yml` file. 
- The `.release/ci.hcl` file contains some metadata for our release service. The only part you'll likely ever need to change is the `release_branches` array -- when you want to release off of a new release branch, update this value to the release branch name and be sure to update the branches array in `.github/workflows/build.yml` to the same value. This should be set to either `main`, the release branch that we'll be releasing off of (once release branches are adopted), or a test branch that you are working off of if you need to validate every part of the pipeline before merging.
- The `netcgo` tag has been added for darwin binaries to avoid issues that customers have faced connecting over a VPN. More detail in this [RFC](https://docs.google.com/document/d/1dKH7n6Zh60drvwDjEGNJf5pcJ_Z_-o7DEag_BufMVx4/edit). After this PR is merged, the next Packer release will close https://github.com/hashicorp/packer/issues/9710  

Things to call out in the changelog for the next release using CRT:

- The [dockerfile](https://github.com/hashicorp/packer/blob/main/Dockerfile) that is used to build the packer docker image available at [hashicorp/packer](https://hub.docker.com/repository/docker/hashicorp/packer) now lives in the root of this repo. The README that describes how to build the docker targets defined in the Dockerfile are available under [./release/docker/README.md](https://github.com/hashicorp/packer/blob/main/.release/docker/README.md).
- Docker images are now available for all supported architectures that the linux binaries are built for including arm, arm64, 386, and amd64
- Packer's linux package service configs and pre/post install scripts are now available under [.release/linux](https://github.com/hashicorp/boundary/blob/main/.release/linux)
- Packer's linux packages are now available for all supported linux architectures including arm, arm64, 386, and amd64

Other odds and ends to note: 
- All CircleCI jobs were ported over to GitHub Actions
- The acceptance tests will now run on a cron on the default branch every day at midnight. In the future, we can add slack notifications to make failures more visible to the team
- After merging and releasing, https://github.com/hashicorp/docker-hub-images can be archived
- After merging and releasing, https://github.com/hashicorp/packer-releases can be archived
- After merging and releasing, we can remove the CircleCI hooks from this project
- Shared [documentation exists here](https://docs.google.com/document/d/14v-KQKhwSsfduJQPaD13GZtHwxIUx0nRDFspKtZAslQ/edit#heading=h.2t5zdr4q7jom) that describes how to promote artifacts to staging and to production.

Still TODO:
- ~Add website-docker-image job secrets to GitHub Actions~
- ~Add acceptance test secrets to GitHub Actions and change the workflow to run on either merges to `master` or at a scheduled time (e.g. nightly)~
- ~Port over nightly github prereleases to GHA~
- ~Sync with Packer team to validate built artifacts and align on naming convention for docker images pushed to DockerHub and AWS ECR~

### CODE REVIEW

File: .github/workflows/acceptance-test.yml
Comment (mdeggies) [inline .github/workflows/acceptance-test.yml:17]: Example CI run of failing tests: https://github.com/hashicorp/packer/runs/5164743184. Need to follow-up to see if these are related to the CI changes or not

File: .github/workflows/build.yml
Comment (samsalisbury) [inline .github/workflows/build.yml:245]: I believe authenticating to Artifactory shouldn't be needed here, as we now [have sign-off from security to reference a public image](https://github.com/hashicorp/actions-docker-build/pull/8#issuecomment-1035075193) with a verified digest. The QEMU changes, including removing Artifactory login and using public image, are now also [merged into main](https://github.com/hashicorp/actions-docker-build/commit/ddaad7b301cc1b2e7e6201302ded2f3be5472812) in `actions-docker-build`, so we should cut a release of that and then remove these lines along with the GitHub secrets I think (unless they are needed elsewhere?)

@smacfarlane would you like to make a new release of `actions-docker-build` today? As long as the change isn't breaking which I don't think it is, you can add a `v1.2` tag for the GitHub Release itself, and also force-move the `v1` tag to point at `v1.2`. If it's a breaking change we'll need to go to `v2` though.

File: Dockerfile
Comment (samsalisbury) [inline Dockerfile:16]: In the doc above, and in the comment just above this line, it says that the `dev` image is the same as `full` in that it contains all the source code. However, this target seems to just use an `ADD` to copy a prebuilt binary inside the image.

File: Dockerfile
Comment (samsalisbury) [inline Dockerfile:103]: As this is the final target defined in the Dockerfile, it will be the default one that gets built when running `docker build`. In helloworld, we have a final line in the Dockerfile which is just `FROM dev` the result of that line is that the `dev` image becomes the default. Just something we might want to use here too for consistency?

**Review by samsalisbury — COMMENTED** [review]
Hi @mdeggies I only reviewed the docker parts, I hope this is helpful, let me know if I can help with addressing any of the issues or improving documentation around them, thanks :)

File: .github/workflows/build.yml
Comment (mdeggies) [reply .github/workflows/build.yml:245]: Thanks for the review, Sam! We have a separate Jira ticket to track removing the login to artifactory / setup-qemu steps across projects when there's new version of the docker-build action available. We had to add across multiple projects to prevent docker arm build failures. If the new version is tagged and available before this PR is merged, I'll push up a commit to remove this.

File: Dockerfile
Comment (mdeggies) [reply Dockerfile:16]: The instructions on building the dev image are in the makefile. I can make these comments consistent so it's clear what's happening. Thanks! :)

File: Dockerfile
Comment (mdeggies) [reply Dockerfile:20]: +1, I'll make this change. I had just ported over the dockerfiles from https://github.com/hashicorp/docker-hub-images with only the necessary changes. There's probably a lot that can be cleaned up in future work.

File: Dockerfile
Comment (mdeggies) [reply Dockerfile:28]: Multiple projects use the target name `official` to refer to images that include 'official' binaries from releases.hashicorp.com. Pre-CRT, all teams released images as a post-release step using these 'official' binaries from the releases site. Here's an example from boundary: https://github.com/hashicorp/boundary/blob/main/Dockerfile#L35. 

The reason this target is included in the dockerfile is that some users want to build images locally using a specific 'official' version of a product, just like they could do pre-CRT. So we definitely need a docker image target to support that.

File: Dockerfile
Comment (mdeggies) [reply Dockerfile:103]: Ah, I didn't know that. I don't think we have that change included in any projects.

File: .github/workflows/acceptance-test.yml
Comment (nywilken) [reply .github/workflows/acceptance-test.yml:17]: Hi @mdeggies the failing don't appear to be related to the ci change. Tehy appear to be related to missing credentials for HCP and AWS. These keys are loaded from secrets directory in the packer-release repo,  which are exposed via environment variables. 

Are these variable that would would have to pull into the Packer's secrets or should they be managed in the common release tooling repo?

**Comment by nywilken** [comment]
> Artifact builds will now only run on merges to the release branches or to main
Are you referring to the Artifact builds that get built from PRs? For example https://app.circleci.com/pipelines/github/hashicorp/packer/11432/workflows/f32c1dc5-9e34-4e48-b267-7d2f6e46b698/jobs/140848

File: .github/workflows/build.yml
Comment (nywilken) [inline .github/workflows/build.yml:191]: Just for context. In v1.7.9 an issue was reported about building Packer on darwin/arm64. This will help us catch these potential issues sooner. I'm so happy for this change 🥳

**Review by nywilken — COMMENTED** [review]
This is looking really good. Thank you for opening up this PR and helping with the onboarding to CRT. I left a couple of questions so far.

 I'll continue reviewing and look to test the built artifacts.

File: Dockerfile
Comment (samsalisbury) [reply Dockerfile:28]: Ah, that's great to know, thank you! Maybe we should add an example to the helloworld projects as well to share this pattern more widely?

Comment (samsalisbury) [reply Dockerfile:28]: Tracked this idea here in case it's useful: https://hashicorp.atlassian.net/browse/RDX-337

File: .github/workflows/acceptance-test.yml
Comment (mdeggies) [reply .github/workflows/acceptance-test.yml:17]: Ahh, thanks for looking into this! For now, we'll need to port over the required secrets from packer-release into github repo level secrets in this project, and set the job to run on only merges to `master` or on a schedule (nightly or whatever makes sense for your team). In the future, this setup will change and tests will run within the common workflows repo.

File: .github/workflows/build.yml
Comment (mdeggies) [reply .github/workflows/build.yml:121]: I so wish that were possible :/ 
I'll create a backlog item to follow up on this with github, but for now, parameterizing inputs fed into build matrices does not work within GHA.

File: Dockerfile
Comment (mdeggies) [inline Dockerfile:108]: `full` is a weird docker target right now, and it's really not what we want.

In CI, the docker build action copies the CI-built binaries into the container under /bin/. E.g. the packer amd64 go binary built in github actions will be included in the docker amd64 image.. we **can** clone the packer repo and build from `master` within the dockerfile for this target, but we will be overwriting whatever has already been copied into /bin/. 

Looking through `scripts/build.sh`, I don't see anything being built other than the go binary for different goos/goarches, so I wonder if we want to use this at all. The docker build step will run when merges happen to `master`, so the binaries built and included in the docker image will always be the latest from `master`.. so my meta question is, do we need/want to continue supporting `full`?

File: Dockerfile
Comment (mdeggies) [reply Dockerfile:31]: Thanks! This is great

File: .release/ci.hcl
Comment (alvin-huang) [inline .release/ci.hcl:6]: This channel ID is incorrect
```suggestion
    notification_channel = "C04SJ33GY"
```

File: .release/ci.hcl
Comment (alvin-huang) [inline .release/ci.hcl:13]: Note to remove this branch before merging like the other references in other files

File: .github/workflows/website-docker-image.yml
Comment (alvin-huang) [inline .github/workflows/website-docker-image.yml:24]: Note to add this secret prior to merging

**Comment by alvin-huang** [comment]
@kpenfound would you mind double checking the Linux packaging files?

File: Makefile
Comment (alvin-huang) [inline Makefile:81]: can this version be parsed with `version.sh` so we don't have to modify it in two places?

File: .release/ci.hcl
Comment (mdeggies) [reply .release/ci.hcl:6]: I'm using this for testing :) I'll update it to #proj-packer-releases before merging.

File: .github/workflows/website-docker-image.yml
Comment (mdeggies) [reply .github/workflows/website-docker-image.yml:24]: Yep, this is called out in the PR notes! Thanks

File: Makefile
Comment (mdeggies) [reply Makefile:81]: It's a great idea! The issue I ran into was while running some of the `make docker-*` commands using the version set in `version.go`. `make docker-official`, for ex, fails because the version is set to the next patch version which hasn't been released yet. I don't have a great solution right now, but we could think of something to benefit other projects as well.

File: .github/workflows/build.yml
Comment (samsalisbury) [reply .github/workflows/build.yml:245]: Hi @mdeggies we released the update to `actions-docker-build` yesterday, so the `v1` tag you're already using now points to `v1.2` which includes this change, along with the smoke test work done by @jeanneryan.

So, with any luck if you remove this code block things should continue to work. If you hit any issues, please let us know straight away! Thanks :)

File: Dockerfile
Comment (mdeggies) [reply Dockerfile]: I'm removing the `full` target for now.

File: .github/workflows/build.yml
Comment (mdeggies) [reply .github/workflows/build.yml]: Note: This is updated in the latest version :-)

File: Makefile
Comment (mdeggies) [reply Makefile:81]: Going to resolve this for now. We can look into refactoring later on.

File: .github/workflows/build.yml
Comment (claire-labry) [inline .github/workflows/build.yml:13]: Do we any needs / use cases for adding any `GO_TAGS`? It can be left as an empty string in case any metadata can be appended Like its done in [CTS](https://github.com/hashicorp/consul-terraform-sync/blob/6049b348b13ad9ed599fb4ad33858f0a2fd722e6/.github/workflows/build.yml#L15) and the GO_TAG being used [here](https://github.com/hashicorp/consul-terraform-sync/blob/6049b348b13ad9ed599fb4ad33858f0a2fd722e6/.github/workflows/build.yml#L168)

Just a thought!

Comment (mdeggies) [reply .github/workflows/build.yml:13]: That's a great idea! Adding this change in :)

File: .github/workflows/build.yml
Comment (claire-labry) [inline .github/workflows/build.yml:113]: @mdeggies I just realized something -- I think the set up Git step here is only for enterprise products. @kpenfound or @alvin-huang please double check me on this.

File: .go-version
Comment (nywilken) [inline .go-version:1]: Do we need to update this file when upgrading to a new version of Go?

File: .release/docker/README.md
Comment (nywilken) [inline .release/docker/README.md:89]: Note to @hc-github-team-packer we need to update this README once https://github.com/hashicorp/packer/issues/11598 is complete

File: CODEOWNERS
Comment (nywilken) [inline CODEOWNERS:74]: I pushed an update to fix the issue. This may cause a conflict here that needs to be pulled in.

File: Makefile
Comment (nywilken) [inline Makefile:206]: 👍

**Review by nywilken — COMMENTED** [review]
This looks really good to me. I left two questions. 

Previously, upon a successful release the CHANGELOG would be updated to reflect the latest release version. Is that handled in the CRT? Or do we need to add the upcoming version information ourselves after a release. It is okay if we do. I'm just asking so that I can update our docs.

File: CODEOWNERS
Comment (mdeggies) [reply CODEOWNERS:3]: This is great. TY!

File: .go-version
Comment (mdeggies) [reply .go-version:1]: Yes! If the version of go changes that is used to build packer, this will need to be updated.

**Comment by mdeggies** [comment]
> Previously, upon a successful release the CHANGELOG would be updated to reflect the latest release version. Is that handled in the CRT? Or do we need to add the upcoming version information ourselves after a release. It is okay if we do. I'm just asking so that I can update our docs.

We don't have this automated in CRT yet, but it's on our short term roadmap, so hopefully we'll be able to get to it soon. We post updates like this over in #feed-crt-newsletter, so feel free to hop in there to subscribe to updates.

**Review by nywilken — APPROVED** [review]
We are going to officially release next week. So we will walk through the whole process. If there are any kinks we can address them in the actual release workflow. Does that sound good to you? 

If so, please feel free to merge when ready.

File: .github/workflows/build.yml
Comment (mdeggies) [reply .github/workflows/build.yml]: Thank you again, Claire! 💯

**Pull request merged**
